### PR TITLE
migrate onboard for az

### PR DESCRIPTION
**Pull request by qiaozha**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### CODE REVIEW

File: src/migrate/azext_migrate/actions.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        super(ImageBuilderAddCustomize, self).__call__(parser, namespace, action, option_string)
+
+    def get_action(self, values, option_string):  # pylint: disable=no-self-use
+        try:
+            properties = dict(x.split('=', 1) for x in values)
+        except ValueError:
+            raise CLIError('usage error: {} [KEY=VALUE ...]'.format(option_string))
+        d = {}
+        for k in properties:
+            kl = k.lower()
```
[the remark this answers is not in the document]
Comment (qiaozha) [reply src/migrate/azext_migrate/actions.py]: probably some mistake, will fix it later

File: src/migrate/azext_migrate/custom.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                          resource_group_name,
+                          project_name):
+    return client.get(resource_group_name=resource_group_name, project_name=project_name)
+
+
+def migrate_projects_create(cmd, client,
+                            resource_group_name,
+                            project_name,
+                            project=None):
+    return client.create(resource_group_name=resource_group_name, project_name=project_name, project=project)
```
[the remark this answers is not in the document]
Comment (qiaozha) [reply src/migrate/azext_migrate/custom.py]: I just change it

File: src/migrate/azext_migrate/actions.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+
+class AddParameters(argparse._AppendAction):
+    def __call__(self, parser, namespace, values, option_string=None):
+        action = self.get_action(values, option_string)
+        super(ImageBuilderAddCustomize, self).__call__(parser, namespace, action, option_string)
+
+    def get_action(self, values, option_string):  # pylint: disable=no-self-use
+        try:
+            properties = dict(x.split('=', 1) for x in values)
```
[the remark this answers is not in the document]
Comment (qiaozha) [reply src/migrate/azext_migrate/actions.py]: Exactly. Will need more work on that.

File: src/migrate/azext_migrate/_help.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    short-summary: migrate assessment-options
+"""
+
+helps['migrate assessment-options show'] = """
+    type: command
+    short-summary: Get the available options for the properties of an assessment.
+    examples:
+      - name: AssessmentOptions_Get
+        text: |-
+               az migrate assessment-options show --location-name "SoutheastAsia"
```
Comment (yungezz) [inline src/migrate/azext_migrate/_help.py]: is the location not the global location?

File: src/migrate/azext_migrate/_params.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    with self.argument_context('migrate projects list') as c:
+        c.argument('resource_group_name', resource_group_name_type)
+
+    with self.argument_context('migrate projects show') as c:
+        c.argument('resource_group_name', resource_group_name_type)
+        c.argument('project_name', id_part=None, help='Name of the Azure Migrate project.')
+
+    with self.argument_context('migrate projects create') as c:
+        c.argument('resource_group_name', resource_group_name_type)
+        c.argument('project_name', id_part=None, help='Name of the Azure Migrate project.')
```
Comment (yungezz) [inline src/migrate/azext_migrate/_params.py]: project is enough, and it should support both name and resource id

File: src/migrate/azext_migrate/_params.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        c.argument('e_tag', id_part=None, help='For optimistic concurrency control.')
+        c.argument('location', arg_type=get_location_type(self.cli_ctx))
+        c.argument('customer_workspace_id', id_part=None, help='ARM ID of the Service Map workspace created by user.')
```
Comment (yungezz) [inline src/migrate/azext_migrate/_params.py]: same as above

File: src/migrate/azext_migrate/_params.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        c.argument('customer_workspace_location', id_part=None, help='Location of the Service Map workspace created by user.')
+        c.argument('provisioning_state', arg_type=get_enum_type(['Accepted', 'Creating', 'Deleting', 'Failed', 'Moving', 'Succeeded']), id_part=None, help='Provisioning state of the project.')
```
Comment (yungezz) [inline src/migrate/azext_migrate/_params.py]: didn't get it. provisioning_state is a readyonly property right? why it's a argument of create.

File: src/migrate/azext_migrate/actions.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        for k in properties:
+            kl = k.lower()
+            v = properties[k]
+        return d
+
+
+class AddProject(argparse._AppendAction):
+    def __call__(self, parser, namespace, values, option_string=None):
+        action = self.get_action(values, option_string)
+        super(ImageBuilderAddCustomize, self).__call__(parser, namespace, action, option_string)
```
Comment (yungezz) [inline src/migrate/azext_migrate/actions.py]: didn't look deep in the code, but `ImageBuilderAddCustomize`?

File: src/migrate/azext_migrate/azext_metadata.json [hunk 9]
```diff
@@ -0,0 +1,5 @@
+{
+    "azext.isPreview": true,
+    "azext.minCliCoreVersion": "2.0.67",
+    "azext.maxCliCoreVersion": "3.0.0"
```
Comment (yungezz) [inline src/migrate/azext_migrate/azext_metadata.json]: not necessary

File: src/migrate/azext_migrate/custom.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+                               scaling_factor,
+                               percentile,
+                               time_range,
+                               stage,
+                               currency,
+                               azure_hybrid_use_benefit,
+                               discount_percentage,
+                               sizing_criterion,
+                               e_tag=None):
+    return client.create(resource_group_name=resource_group_name, project_name=project_name, group_name=group_name, assessment_name=assessment_name, e_tag=e_tag, azure_location=azure_location, azure_offer_code=azure_offer_code, azure_pricing_tier=azure_pricing_tier, azure_storage_redundancy=azure_storage_redundancy, scaling_factor=scaling_factor, percentile=percentile, time_range=time_range, stage=stage, currency=currency, azure_hybrid_use_benefit=azure_hybrid_use_benefit, discount_percentage=discount_percentage, sizing_criterion=sizing_criterion)
```
Comment (yungezz) [inline src/migrate/azext_migrate/custom.py]: how about break it into multiple lines for readability?

File: src/migrate/azext_migrate/actions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (qiaozha) [reply src/migrate/azext_migrate/actions.py]: This action part should have been fixed in the latest code

File: src/migrate/azext_migrate/actions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (qiaozha) [reply src/migrate/azext_migrate/actions.py]: This is still in the process

File: src/migrate/azext_migrate/actions.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    def get_action(self, values, option_string):  # pylint: disable=no-self-use
+        try:
+            properties = dict(x.split('=', 1) for x in values)
+        except ValueError:
+            raise CLIError('usage error: {} [KEY=VALUE ...]'.format(option_string))
+        d = {}
+        for k in properties:
+            kl = k.lower()
+            v = properties[k]
+        return d
```
[the remark this answers is not in the document]
Comment (qiaozha) [reply src/migrate/azext_migrate/actions.py]: yeah, the problem should have been fixed

File: src/migrate/azext_migrate/_help.py (hunk 4 above)
Comment (qiaozha) [reply src/migrate/azext_migrate/_help.py]: Yes. it is. This should have been changed into location but I am wondering how about parameters like location-id. Should it also be changed into location?

File: src/migrate/azext_migrate/_params.py (hunk 5 above)
Comment (qiaozha) [reply src/migrate/azext_migrate/_params.py]: Are all the resource references support that kind of feature?

File: src/migrate/azext_migrate/_params.py (hunk 7 above)
Comment (qiaozha) [reply src/migrate/azext_migrate/_params.py]: This should have been removed in the latest code

File: src/migrate/azext_migrate/actions.py (hunk 8 above)
Comment (qiaozha) [reply src/migrate/azext_migrate/actions.py]: This should have been fixed in the latest code generator

File: src/migrate/azext_migrate/azext_metadata.json (hunk 9 above)
Comment (qiaozha) [reply src/migrate/azext_migrate/azext_metadata.json]: This should have been fixed in the latest code generator

File: src/migrate/azext_migrate/custom.py (hunk 10 above)
Comment (qiaozha) [reply src/migrate/azext_migrate/custom.py]: This line too long is in the process of fixing. I think it will be fixed very soon.

**Comment by MyronFanQiu** [comment]
closed it after discussion with @qiaozha

**Pull request closed without merging**